File: server/fastapi/config/routes/30_statsig.routes.py
"""GET /healthz/integrations/statsig/gates — proxy of upstream /console/v1/gates."""

from __future__ import annotations

from typing import TYPE_CHECKING

from app_yaml_fetch_config import get_config
from fastapi import APIRouter, Depends, Request, Response, status
from fetch_http_client import AsyncClient

from ._di import get_runtime_template_resolver, get_statsig_client
from ._provider_config_echo import build_echo

if TYPE_CHECKING:
    from ..lifecycles._runtime_template_resolver import ResolverHandle


def mount(_app, _config) -> APIRouter:
    router = APIRouter()

    @router.get("/healthz/integrations/statsig/gates")
    async def statsig_feature_gates(
        request: Request,
        response: Response,
        client: AsyncClient = Depends(get_statsig_client),
        resolver: "ResolverHandle" = Depends(get_runtime_template_resolver),
    ):
        cfg = get_config()
        host = (cfg.get("providers") or {}).get("statsig", {}).get("base_url", "")
        try:
            config_used = await build_echo(
                "statsig", request, cfg, resolver, trigger="OnRequest"
            )
        except Exception:  # noqa: BLE001
            config_used = None
        try:
            upstream = await client.get("/gates")
        except Exception as exc:  # noqa: BLE001
            response.status_code = status.HTTP_502_BAD_GATEWAY
            return {
                "service": "statsig",
                "connected": False,
                "error": str(exc),
                "config_used": config_used,
            }
        if upstream.status_code >= 400:
            response.status_code = status.HTTP_502_BAD_GATEWAY
            return {
                "service": "statsig",
                "connected": False,
                "upstream_status": upstream.status_code,
                "upstream_body": upstream.text[:512] if hasattr(upstream, "text") else "",
                "config_used": config_used,
            }
        body = upstream.json()
        rows = (body.get("data") or [])[:5] if isinstance(body, dict) else []
        gates = [
            {"id": r.get("id"), "name": r.get("name"), "isEnabled": r.get("isEnabled")}
            for r in rows
        ]
        return {
            "service": "statsig",
            "connected": True,
            "host": host,
            "count": len(gates),
            "gates": gates,
            "config_used": config_used,
        }

    return router

File: server/fastify/config/routes/30_statsig.routes.mjs
import { getConfig } from "@ployglot/app-yaml-fetch-config";
import { buildEcho } from "../lifecycles/_provider_config_echo.mjs";

export default async function statsigRoutes(fastify, _config) {
  fastify.get("/healthz/integrations/statsig/gates", async (request, reply) => {
    const cfg = getConfig();
    const host = cfg?.providers?.statsig?.base_url ?? "";
    let config_used = null;
    try {
      config_used = await buildEcho({
        provider: "statsig",
        request,
        cfg,
        resolver: request.runtime_template_resolver,
        trigger: "OnRequest",
      });
    } catch {
      config_used = null;
    }
    const client = await request.fetchClient("statsig");
    try {
      const upstream = await client.get("/gates");
      if (upstream.status_code >= 400) {
        const body = await upstream.text();
        return reply.code(502).send({
          service: "statsig",
          connected: false,
          upstream_status: upstream.status_code,
          upstream_body: body.slice(0, 512),
          config_used,
        });
      }
      const body = await upstream.json();
      const gates = (body?.data ?? []).slice(0, 5).map((g) => ({
        id: g.id,
        name: g.name,
        isEnabled: g.isEnabled,
      }));
      return {
        service: "statsig",
        connected: true,
        host,
        count: gates.length,
        gates,
        config_used,
      };
    } catch (err) {
      return reply.code(502).send({
        service: "statsig",
        connected: false,
        error: err.message,
        config_used,
      });
    }
  });
}
